### DISCUSSION TITLE
How to use `Readline`?

### DISCUSSION THREAD

**Discussion by williamrjribeiro** [TheSpyder/rescript-nodejs#6] [Q&A] [answered]
Hi there,

I just started learning ReScript and I would like to use it for a NodeJS application. I'm trying to port this piece of NodeJS code to ReScript using `rescript-nodejs` but I can't find a way to use `Readline` module:

```js
const fs = require('fs');
const readline = require('readline');

const rl = readline.createInterface({
  input: fs.createReadStream('sample.txt'),
  crlfDelay: Infinity
});

rl.on('line', (line) => {
  console.log(`Line from file: ${line}`);
});
```

I couldn't find any documentation on `Readline` module. Can you give me some hints? I just want to read a  text file line by line.

If this is not the right place, where should I ask this kind of question?
Thanks.

**Comment by TheSpyder** [comment]
Hi! This is a fine place to ask. The [ReScript forums](https://forum.rescript-lang.org) are also good but people there might just tell you to write your own bindings 🙂 

I'll see if I can port your code to use this library and post the results later today (I'm in Australia). I might also enable github discussions for the repo and transfer your question there.

**Comment by TheSpyder** [comment]
Oh wow, looks like the Readline module was written incorrectly. The `make` function requires a `t`, but the only way to create a `t` is with the `make` function:
https://github.com/TheSpyder/rescript-nodejs/blob/83e9ebafa898f888f8c0e6e35949c56607823146/src/Readline.res#L17

It also seems weird that the whole file is inside an `Interface` sub-module.

I'll look up the nodejs documentation and make sure I get the module right, then push a new release and show you how to use it.

**Comment by TheSpyder** [comment, marked as answer]
Published as 14.2.0 (it's a breaking change to `Readline`, but the module wasn't usable before so I figured a minor was ok).

The sample code you requested:
https://github.com/TheSpyder/rescript-nodejs/blob/6214ab63ca7644fb10c8f17a7b97b3879afbbe66/examples/ReadFileByLine.res#L4-L12

Which generates the correct JS (`Pervasives.infinity` is just a reference to `Infinity`):
https://github.com/TheSpyder/rescript-nodejs/blob/6214ab63ca7644fb10c8f17a7b97b3879afbbe66/lib/js/examples/ReadFileByLine.bs.js#L4-L16

I'm happy to discuss what each part of the code does if you need help with it.

**Reply by williamrjribeiro** [reply]
Oh wow! I was not expecting this quickness! You sir deserve at least a nice beer for all this work. Thank you very much! 🍻 

The ReScript code is easy to follow (as I was expecting) I just had never seen/heard `Pervasives` before. 👍

**Reply by TheSpyder** [reply]
Happy to help 😁 I want this library to be useful, having a broken module is really bad.

`Pervasives` is one of those OCaml concepts that the ReScript team don't want anyone to know about and may one day be replaced by a module of their own. It's the central (pervasive) module of the OCaml standard library that is available when no `open` statements are used. Every module has an implicit `open Pervasives` at the top.

I believe in later OCaml versions the implicit open was changed to `Standard`, but that isn't really relevant to ReScript.
